dataclients/kubernetes/path_test.go

```diff
@@ -0,0 +1,392 @@
+	for _, ri := range r {
+		if len(ri.PathRegexps) == 1 {
+			return ri
+		}
+	}
+
+	return nil
+}
+
+func findRouteWithPathPrefix(r []*eskip.Route) (*eskip.Route, error) {
+	for _, ri := range r {
+		p, err := findPathPredicate(ri, "PathSubtree")
+		if err != nil {
+			return nil, err
+		}
+
+		if p != nil {
+			return ri, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func findRouteWithExactPath(r []*eskip.Route) (*eskip.Route, error) {
+	for _, ri := range r {
+		p, err := findPathPredicate(ri, "Path")
+		if err != nil {
+			return nil, err
+		}
+
+		if p != nil && ri.Path != "" {
+			return nil, errors.New("route with duplicate path predicate found")
+		}
+
+		if p != nil || ri.Path != "" {
+			return ri, nil
+		}
+	}
+
+	return nil, nil
+}
+
+func TestPathMatchingModes(t *testing.T) {
+	s := testServices()
+	api := newTestAPI(t, s, &ingressList{})
+	defer api.Close()
+
+	setIngressWithPath := func(p string, annotations ...string) {
+		i := testIngress(
+			"namespace1", "ingress1", "service1", "", "", "", "", backendPort{8080}, 1.0,
+			testRule("www.example.org", testPathRule(p, "service1", backendPort{8080})),
+		)
+
+		annotation := strings.Join(annotations, " && ")
+		if len(annotations) > 0 {
+			i.Metadata.Annotations[skipperpredicateAnnotationKey] = annotation
+		}
+
+		api.ingresses.Items = []*ingressItem{i}
+	}
+
+	loadRoutes := func(m PathMode) ([]*eskip.Route, error) {
+		c, err := New(Options{
+			KubernetesURL: api.server.URL,
+			PathMode:      m,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		defer c.Close()
+		return c.LoadAll()
+	}
+
+	t.Run("default", func(t *testing.T) {
+		setIngressWithPath("/foo")
+		r, err := loadRoutes(KubernetesIngressMode)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		routeWithRx := findRouteWithRx(r)
+		if routeWithRx == nil {
+			t.Fatal("route with path regexp not found")
+		}
+
+		if routeWithRx.PathRegexps[0] != "^/foo" {
+			t.Error("invalid path regexp value", routeWithRx.PathRegexps[0])
+		}
+	})
+
+	t.Run("regexp", func(t *testing.T) {
+		setIngressWithPath("^/foo")
+		r, err := loadRoutes(PathRegexp)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		routeWithRx := findRouteWithRx(r)
+		if routeWithRx == nil {
+			t.Fatal("route with path regexp not found")
+		}
+
+		if routeWithRx.PathRegexps[0] != "^/foo" {
+			t.Error("invalid path regexp value", routeWithRx.PathRegexps[0])
+		}
+	})
+
+	t.Run("path prefix", func(t *testing.T) {
+		setIngressWithPath("/foo")
+		r, err := loadRoutes(PathPrefix)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		routeWithPathPrefix, err := findRouteWithPathPrefix(r)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if routeWithPathPrefix == nil {
+			t.Fatal("route with path prefix not found")
+		}
+
+		p, err := findPathPredicate(routeWithPathPrefix, "PathSubtree")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if p.Args[0] != "/foo" {
+			t.Error("invalid path prefix value", p.Args[0])
+		}
+	})
+
+	t.Run("additional exact path from annotation", func(t *testing.T) {
```

Comment:
this does not have an implementation anymore

Reply:
currently we don't prevent somebody adding an path predicate (Path() or PathSubtree()) via annotation, and I think this is fine. With these tests we just verify that both regexp based path and the predicate path is there, because when use path prefix mode, only one of the should be kept.